### PR TITLE
Linux and Windows Support

### PR DESCRIPTION
**Pull request by jshier**
### Issue Link :link:
#1935

### Goals :soccer:
This PR builds on #3098 and #3462 to add preliminary support for Alamofire on Linux and Windows now that Swift 5.4 greatly decreases the changes needed between the platforms.

### Implementation Details :construction:
Functionally, this PR doesn't change implementation except where  types are unavailable on non-Apple. This includes:

- `AFError` support for `ServerTrustFailureReason` has been checked.
- All Security framework type usage has been checked.
- Various `UT` types from CoreServices were checked.
- Access to various Info.plist values has been moved from CoreFoundation constants to raw `String`s.
- Missing APIs were checked.

### TODO ✅ 

- [x] Enable `URLAuthenticationChallenge` for everything that's not `ServerTrust`.
  - Enabled, but doesn't work on Linux / Windows.
- [ ] Replace `getBoundStreams` usage.
  - Needs solution.
- [ ] Replace MIME type detection for `MultipartFormData`.
  - Needs solution.

### Testing Details :mag:
Testing needs to be updated to run against Firewalk.

### CODE REVIEW

**Comment by alex-taffe** [comment]
I’d like to get Windows support in here. Thoughts on renaming this “Linux/Windows support” and combining the efforts from #3462?

**Comment by jshier** [comment]
@alex-taffe Sure. Once you're done with the PR we can merge it in here and work towards full testing support.

**Comment by alex-taffe** [comment]
Awesome, will let you know when it’s done, should probably just mark as a draft

**Comment by alex-taffe** [comment]
Actually @jshier, the local build just completed on my Windows box and I think we’re good to go.

**Comment by jshier** [comment]
@alex-taffe If you're interested, our next step would be to get our [Firewalk](https://github.com/alamofire/Firewalk) test server building and running on Linux and Windows so we can start executing tests. This includes some sort of deployment mechanism for both platforms, whether as a raw binary or through a package manager (Homebrew on Linux?). It's a Vapor server so it should already be cross platform, so it's mainly a tooling issue. Would you be interested in helping with that?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -128,3 +128,43 @@ jobs:
...
+        Copy-Item "$env:SDKROOT\usr\share\visualc.modulemap" -destination "$env:VCToolsInstallDir\include\module.modulemap"
+        Copy-Item "$env:SDKROOT\usr\share\visualc.apinotes" -destination "$env:VCToolsInstallDir\include\visualc.apinotes"
+        Copy-Item "$env:SDKROOT\usr\share\winsdk.modulemap" -destination "$env:UniversalCRTSdkDir\Include\$env:UCRTVersion\um\module.modulemap"
+    - name: SPM Windows build
+      shell: cmd
+      run: |
+        cd ${{ github.workspace}}
+        set SDKROOT=%SystemDrive%\Library\Developer\Platforms\Windows.platform\Developer\SDKs\Windows.sdk
+        %SystemDrive%\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\bin\swift-build.exe -c debug -Xlinker /INCREMENTAL:NO -v
+        if not exist .build\x86_64-unknown-windows-msvc\debug\alamofire.exe exit 1
```
[the remark this answers is not in the document]
Comment (jshier) [reply .github/workflows/ci.yml]: Looks like that fixed it, nice!

**Comment by alex-taffe** [comment]
> @alex-taffe If you're interested, our next step would be to get our [Firewalk](https://github.com/alamofire/Firewalk) test server building and running on Linux and Windows so we can start executing tests. This includes some sort of deployment mechanism for both platforms, whether as a raw binary or through a package manager (Homebrew on Linux?). It's a Vapor server so it should already be cross platform, so it's mainly a tooling issue. Would you be interested in helping with that?

I would love to, but unfortunately Vapor doesn’t work on Windows yet. I had tried using it a few months ago for a project,  but Vapor depends on SwiftNIO (see https://github.com/vapor/vapor/blob/9d154df6214a1d6027f04a495e8981bac49224da/Package.swift#L19). compnerd (not tagging just to avoid junking up his inbox) is working on SwiftNIO support on Windows, but it’s not done yet (https://github.com/apple/swift-nio/issues/1673)

**Comment by alex-taffe** [comment]
> @alex-taffe Since SPM is supposed to work on Windows now, would it be possible to update the Windows build to use that instead of a direct executable?

I’m not entirely sure what you mean. Do you mean replacing `%SystemDrive%\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\bin\swift-build.exe` with something more simplistic? SPM is indeed supported (I’m actually using Alamofire as an SPM dependency in an additional project on Windows). We just need that massive command on GitHub Actions because certain environment variables aren’t set properly by the installer on the runner. On my local Windows box, you can just run `swift build`.

**Comment by jshier** [comment]
> I’m not entirely sure what you mean. Do you mean replacing `%SystemDrive%\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\bin\swift-build.exe` with something more simplistic? SPM is indeed supported (I’m actually using Alamofire as an SPM dependency in an additional project on Windows). We just need that massive command on GitHub Actions because certain environment variables aren’t set properly by the installer on the runner. On my local Windows box, you can just run `swift build`.

I just meant that on Linux we can build (and eventually test) the framework with just `swift build`. If we need the more complex command that's fine.

> I would love to, but unfortunately Vapor doesn’t work on Windows yet. I had tried using it a few months ago for a project, but Vapor depends on SwiftNIO (see https://github.com/vapor/vapor/blob/9d154df6214a1d6027f04a495e8981bac49224da/Package.swift#L19). compnerd (not tagging just to avoid junking up his inbox) is working on SwiftNIO support on Windows, but it’s not done yet ([apple/swift-nio#1673](https://github.com/apple/swift-nio/issues/1673))

That's too bad, I was hoping to bring everything into parity. I suppose we need a way to say that Windows just isn't fully supported, which would also apply to Linux until I can get Firewalk ported. You wouldn't have any interest in that, would you?

**Comment by alex-taffe** [comment]
I might be able to take a look this weekend, otherwise it'll probably be in a few weeks. The Windows support was mainly for a project at work that needs to get done soon. Won't fully commit to it, but I'll try to get to it at some point

**Comment by jshier** [comment]
@alex-taffe I've updated the PR so that the test suite builds on Linux and Windows. However, there are rather extensive issues with getting it to pass (authentication crashes, bizarre test failures) even once I got Firewalk running on Linux. So while I think we're okay shipping the changes required to build on the other platforms, Alamofire will be completely unsupported on Linux and Windows.

**Comment by alex-taffe** [comment]
I think I'm good with that

**Pull request merged**